File: 0-event-ingestion-kafka/ingest-api/2-csharp/KafkaConfig.cs
namespace IngestApi;

/// <summary>
/// Strongly typed Kafka settings. Every default is committed so the service runs
/// out-of-the-box under Docker Compose; Compose only overrides the broker list.
/// </summary>
public sealed class KafkaConfig
{
    /// <summary>Comma-separated broker list, e.g. <c>kafka:9092</c>.</summary>
    public string BootstrapServers { get; init; } = "localhost:9092";

    /// <summary>Topic that ad-click events are appended to.</summary>
    public string Topic { get; init; } = "ad-clicks";

    /// <summary>Stable client id reported to the broker for this producer.</summary>
    public string ClientId { get; init; } = "ingest-api";

    /// <summary>Reads the config block from environment variables.</summary>
    public static KafkaConfig FromEnv() => new()
    {
        BootstrapServers = Environment.GetEnvironmentVariable("KAFKA_BOOTSTRAP_SERVERS") ?? "localhost:9092",
        Topic = Environment.GetEnvironmentVariable("KAFKA_TOPIC") ?? "ad-clicks",
        ClientId = Environment.GetEnvironmentVariable("KAFKA_CLIENT_ID") ?? "ingest-api",
    };
}

File: 0-event-ingestion-kafka/ingest-api/0-typescript/src/config/kafka.config.ts
import { registerAs } from "@nestjs/config"

/**
 * Strongly typed shape of the Kafka configuration block.
 *
 * Centralising the type lets every consumer of `ConfigService` read the block
 * with `getOrThrow<KafkaConfig>("kafka")` and get compile-time field checks.
 */
export interface KafkaConfig {
    /** Comma-separated broker list, e.g. `kafka:9092`. */
    brokers: string[]
    /** Topic that ad-click events are appended to. */
    topic: string
    /** Stable client id reported to the broker for this producer. */
    clientId: string
}

/**
 * Registers the `kafka` configuration namespace from environment variables.
 *
 * Every default is committed so the service runs out-of-the-box under Docker
 * Compose without any `.env` file; Compose only overrides the broker address.
 */
export default registerAs<KafkaConfig>("kafka", () => ({
    // Split on comma so multiple brokers can be passed via one env var.
    brokers: (process.env.KAFKA_BOOTSTRAP_SERVERS ?? "localhost:9092").split(","),
    topic: process.env.KAFKA_TOPIC ?? "ad-clicks",
    clientId: process.env.KAFKA_CLIENT_ID ?? "ingest-api",
}))
